### PR TITLE
MLIBZ-1167: Geolocation Queries

### PR DESCRIPTION
**Pull request by heyzooi**
#### Description

Solves the problem of perform geolocation queries in the backend and locally

#### Changes

- NSPredicate -> MongoDB adaptor
- Local cache

#### Tests

- Test geolocation queries with circles with a radio
- Test geolocation queries with polygons (which includes the use case of boxes / rectangles)

### CODE REVIEW

File: Kinvey/Kinvey/Geolocation.swift [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    dynamic var longitude: CLLocationDegrees = 0.0
+    
+    public static func objectForMapping(map: Map) -> BaseMappable? {
+        guard let _: Double = map["latitude"].value(), let _: Double = map["longitude"].value() else {
+            return nil
+        }
+        return GeoPoint()
+    }
+    
+    public convenience init(latitude: Double, longitude: Double) {
```
Comment (tejasranade) [inline Kinvey/Kinvey/Geolocation.swift]: @heyzooi if we have the constructor for `CLLocation`, what is the reason to have this one separately? I prefer we keep one way of doing things when possible.

File: Kinvey/Kinvey/Geolocation.swift [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        self.longitude = coordinate.longitude
+    }
+    
+    public convenience init(location: CLLocation) {
+        self.init()
+        self.latitude = location.coordinate.latitude
+        self.longitude = location.coordinate.longitude
+    }
+    
+    public convenience init(_ array: [Double]) {
```
Comment (tejasranade) [inline Kinvey/Kinvey/Geolocation.swift]: Same comment as above. Exposing this constructor adds unnecessary work e.g. we have to document which value in the array is lat and which is long, etc.

File: Kinvey/Kinvey/Persistable.swift [hunk 3]
```diff
@@ -55,19 +56,19 @@ public func <- <T>(left: inout T!, right: (String, Map)) {
 }
 
 /// Override operator used during the `propertyMapping(_:)` method.
-public func <- <T: Mappable>(left: inout T, right: (String, Map)) {
+public func <- <T: BaseMappable>(left: inout T, right: (String, Map)) {
```
Comment (tejasranade) [inline Kinvey/Kinvey/Persistable.swift]: @heyzooi what is the reason for this change?

File: Makefile [hunk 4]
```diff
@@ -24,7 +24,7 @@ test: test-ios
 
 	
 test-ios:
-	xcodebuild -workspace Kinvey.xcworkspace -scheme Kinvey -destination 'platform=iOS Simulator,name=iPhone 7,OS=10.2' -enableCodeCoverage YES test
+	xcodebuild -workspace Kinvey.xcworkspace -scheme Kinvey -destination 'platform=iOS Simulator,name=iPhone 7,OS=10.1' -enableCodeCoverage YES test
```
Comment (tejasranade) [inline Makefile]: @heyzooi why was this moved down?

File: Kinvey/Kinvey/RealmCache.swift [hunk 5]
```diff
@@ -123,7 +154,55 @@ internal class RealmCache<T: Persistable>: Cache<T> where T: NSObject {
...
     func detach(_ results: RealmSwift.Results<Entity>, query: Query?) -> [T] {
-        return detach(results.map { $0 as! T }, query: query)
+        var results: [T] = results.map { $0 as! T }
+        if let predicate = query?.predicate as? NSComparisonPredicate,
+            let keyPathConstantValuePair = keyPathConstantValuePair(predicate: predicate),
+            let constantValue = keyPathConstantValuePair.constantValueExpression.constantValue,
+            constantValue is MKCircle || constantValue is MKPolygon
+        {
+            if let circle = constantValue as? MKCircle {
+                let center = CLLocation(latitude: circle.coordinate.latitude, longitude: circle.coordinate.longitude)
```
Comment (tejasranade) [inline Kinvey/Kinvey/RealmCache.swift]: I don't full comprehend this change. Let's discuss. On first glance, looks odd in the `RealmCache` class.

File: Kinvey/Kinvey/Geolocation.swift [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+//  Copyright © 2017 Kinvey. All rights reserved.
+//
+
+import Foundation
+import RealmSwift
+import ObjectMapper
+import CoreLocation
+import MapKit
+
+open class GeoPoint: Object, StaticMappable {
```
Comment (heyzooi) [inline Kinvey/Kinvey/Geolocation.swift]: for this one, not really... agree, changing now

File: Kinvey/Kinvey/Geolocation.swift (hunk 1 above)
Comment (heyzooi) [inline Kinvey/Kinvey/Geolocation.swift]: forgot to summarize the constructors, doing it now... that's why I created this kinda agnostic constructor

File: Kinvey/Kinvey/Geolocation.swift (hunk 2 above)
Comment (heyzooi) [inline Kinvey/Kinvey/Geolocation.swift]: same reason... please check the next commit

File: Kinvey/Kinvey/Persistable.swift (hunk 3 above)
Comment (heyzooi) [inline Kinvey/Kinvey/Persistable.swift]: to also cover `StaticMappable` which also extends from `BaseMappable`

File: Makefile (hunk 4 above)
Comment (heyzooi) [inline Makefile]: this is to fix a problem in Travis CI... don't know why, but Travis CI has 2 iOS Simulator with 10.2 and for that reason the script ask you to specify the UUID of the Simulator (which can also change from time to time)

File: Kinvey/Kinvey/RealmCache.swift (hunk 5 above)
Comment (heyzooi) [inline Kinvey/Kinvey/RealmCache.swift]: sure... but basically Realm does not support geolocation queries, so we need to do it manually after fetch the results

File: Makefile (hunk 4 above)
Comment (heyzooi) [inline Makefile]: here's an example: https://travis-ci.org/Kinvey/ios-library/builds/198941783
